cognite_toolkit/cdf_tk/templates/data_classes/_project_directory.py

```diff
@@ -0,0 +1,298 @@
+            None,
+        )
+        if version_hash is None:
+            print(f"Failed to find migration from version '{self._cognite_module_version}'.")
+            exit(1)
+        current_hash = calculate_directory_hash(self.project_dir / COGNITE_MODULES)
+        self._has_changed_cognite_modules = current_hash != version_hash
+        self._changes = changes.slice_from(self._cognite_module_version).as_one_change()
+
+    def create_project_directory(self, clean: bool) -> None:
+        if self.project_dir.exists():
+            print(f"[bold]Upgrading directory {self.target_dir_display}...[/b]")
+        else:
+            print(f"Found no directory {self.target_dir_display} to upgrade.")
+            exit(1)
+
+    def do_backup(self, no_backup: bool, verbose: bool) -> None:
+        if not no_backup and self._has_changed_cognite_modules:
+            print(
+                "[bold yellow]WARNING:[/] The cognite_modules have changed, it will not be upgraded.\n"
+                f"  No backup {'would have been' if self._dry_run else 'will be'} done."
+            )
+            return
+        elif self._has_changed_cognite_modules:
+            print("[bold yellow]WARNING:[/] The cognite_modules have changed, it will not be upgraded.")
+
+        if not no_backup:
+            prefix = "Would have backed up" if self._dry_run else "Backing up"
+            if verbose:
+                print(f"{prefix} {self.target_dir_display}")
+            if not self._dry_run:
+                backup_dir = tempfile.mkdtemp(prefix=f"{self.project_dir.name}.", suffix=".bck", dir=Path.cwd())
+                shutil.copytree(self.project_dir, Path(backup_dir), dirs_exist_ok=True)
+        else:
+            print(
+                "[bold yellow]WARNING:[/] --no-backup is specified, no backup "
+                f"{'would have been' if self._dry_run else 'will be'} done."
+            )
+
+    def print_what_to_copy(self) -> None:
+        if not self._has_changed_cognite_modules:
+            print("  Will upgrade modules and files in place.")
+            super().print_what_to_copy()
+
+    def copy(self, verbose: bool) -> None:
+        if self._has_changed_cognite_modules:
+            return
+        else:
+            super().copy(verbose)
+
+    def set_source(self, git_branch: str | None) -> None:
+        if git_branch is None:
+            return
+
+        self._source = self._download_templates(git_branch, self._dry_run)
+
+    def upsert_config_yamls(self, clean: bool) -> None:
+        if clean:
+            super().upsert_config_yamls(clean)
+            return
+        if self._has_changed_cognite_modules:
+            return
+
+        existing_environments = list(self.project_dir.glob("config.*.yaml"))
+        if len(existing_environments) == 0:
+            print("  [bold yellow]WARNING:[/] No existing config.[env].yaml files found, creating from the defaults.")
```

Comment:
Upgrading an untouched project folder from beta4, I got `  WARNING: No existing config..yaml files found, creating from the defaults.`

Reply:
I struggle to recreate this one. Can you tell me which steps you went through?

I tried

1.  `python -m venv venv`
1.  `venv/Scripts/activate`
1.  `pip install cognite-toolkit==0.1.0b4`
2. `cdf-tk init my_project` 
3. `pip install git+https://github.com/cognitedata/cdf-project-templates.git@create-migration`
4. `cdf-tk init my_project --upgrade`